tests/unit/test_hnsw_tiered.cpp

```diff
@@ -1899,9 +1829,288 @@ TYPED_TEST(HNSWTieredIndexTest, alternateInsertDeleteAsync) {
+        size_t pending_swap_jobs = tiered_index->idToSwapJob.size();
+        EXPECT_EQ(tiered_index->deleteVector(0), 0);
+        if (pending_swap_jobs > maxSwapJobs) {
+            ASSERT_EQ(tiered_index->idToSwapJob.size(), 0);
+            EXPECT_EQ(tiered_index->indexSize(), 0);
+        } else {
+            ASSERT_LE(tiered_index->idToSwapJob.size(), maxSwapJobs);
+            ASSERT_EQ(tiered_index->idToSwapJob.size(), tiered_index->indexSize());
         }
+        EXPECT_EQ(tiered_index->getHNSWIndex()->getNumMarkedDeleted(), tiered_index->indexSize());
     }
 }
+
+TYPED_TEST(HNSWTieredIndexTest, alternateInsertDeleteAsync) {
+    std::shared_ptr<VecSimAllocator> allocator = VecSimAllocator::newVecsimAllocator();
+
+    // Create TieredHNSW index instance with a mock queue.
+    size_t dim = 16;
+    size_t n = 1000;
+    for (size_t maxSwapJobs : {(int)n + 1, 10, 1}) {
+        for (size_t M : {2, 16}) {
+            HNSWParams params = {.type = TypeParam::get_index_type(),
+                                 .dim = dim,
+                                 .metric = VecSimMetric_L2,
+                                 .multi = TypeParam::isMulti(),
+                                 .blockSize = 100,
+                                 .M = M};
+            VecSimParams hnsw_params = CreateParams(params);
+            auto jobQ = JobQueue();
+            auto index_ctx = IndexExtCtx();
+            size_t memory_ctx = 0;
+            TieredIndexParams tiered_params = {.jobQueue = &jobQ,
+                                               .jobQueueCtx = &index_ctx,
+                                               .submitCb = submit_callback,
+                                               .memoryCtx = &memory_ctx,
+                                               .UpdateMemCb = update_mem_callback,
+                                               .primaryIndexParams = &hnsw_params};
+
+            TieredHNSWParams tiered_hnsw_params = {.tieredIndexParams = tiered_params,
+                                                   .maxSwapJobs = maxSwapJobs};
+            auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+                TieredFactory::TieredHNSWFactory::NewIndex(&tiered_hnsw_params, allocator));
+            // Set the created tiered index in the index external context.
+            index_ctx.index_strong_ref.reset(tiered_index);
+
+            size_t per_label = TypeParam::isMulti() ? 5 : 1;
+            size_t n_labels = n / per_label;
+
+            // Launch the BG threads loop that takes jobs from the queue and executes them.
+            bool run_thread = true;
+            for (size_t i = 0; i < THREAD_POOL_SIZE; i++) {
+                thread_pool.emplace_back(thread_main_loop, std::ref(jobQ), std::ref(run_thread));
+            }
+
+            // Create and insert 10 vectors, then delete them right after.
+            size_t batch_size = 10;
```

Comment:
Actually, we do get add in parallel with delete because ingesting to HNSW is done asynchronously, and it takes more time than it takes to insert the vector to the flat buffer. So once we get to the delete part, the vectors move from the flat buffer into HNSW in the background.